cpg-language-cxx/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cxx/CLanguage.kt

```diff
@@ -161,4 +161,63 @@ open class CLanguage :
             )
         return resolveWithImplicitCast(call, initialInvocationCandidates)
     }
+
+    override fun isDerivedFrom(
+        type: Type,
+        targetType: Type,
+        hint: HasType?,
+        targetHint: HasType?
+    ): Boolean {
+        val match = super.isDerivedFrom(type, targetType, hint, targetHint)
+        if (match) {
+            return true
+        }
+
+        // As a special rule, a non-nested pointer and array of the same type are completely
+        // interchangeable
+        if (type.root == targetType.root && type is PointerType && targetType is PointerType) {
+            return true
+        }
+
+        return false
+    }
+
+    override fun tryCast(
+        type: Type,
+        targetType: Type,
+        hint: HasType?,
+        targetHint: HasType?
+    ): CastResult {
+        val match = super.tryCast(type, targetType, hint, targetHint)
+        if (match != CastNotPossible) {
+            return match
+        }
+
+        // Numeric types can be cast implicitly
+        if (type is NumericType && targetType is NumericType) {
+            return ImplicitCastNeeded
```

Comment:
Double-check spec if all numbers are on the same hierarchy level

Reply:
It seems so, because if you have something like this:
```cpp
void f(int x) {}
void f(float x) {}

int main() {
   f(2.0); // 2.0 is a double
}
```

Then clang says its ambiguous, so it seems that double to float has the same ranking as double to int.